coderd/httpmw/apikey.go

```diff
@@ -446,6 +462,207 @@ func ExtractAPIKey(rw http.ResponseWriter, r *http.Request, cfg ExtractAPIKeyCon
 	return key, &actor, true
 }
 
+// validateOAuth2ProviderAppTokenAudience validates that an OAuth2 provider app token
+// is being used with the correct audience/resource server (RFC 8707).
+func validateOAuth2ProviderAppTokenAudience(ctx context.Context, db database.Store, key database.APIKey, r *http.Request) error {
+	// Get the OAuth2 provider app token to check its audience
+	//nolint:gocritic // System needs to access token for audience validation
+	token, err := db.GetOAuth2ProviderAppTokenByAPIKeyID(dbauthz.AsSystemRestricted(ctx), key.ID)
+	if err != nil {
+		return xerrors.Errorf("failed to get OAuth2 token: %w", err)
+	}
+
+	// If no audience is set, allow the request (for backward compatibility)
+	if !token.Audience.Valid || token.Audience.String == "" {
+		return nil
+	}
+
+	// Extract the expected audience from the request
+	expectedAudience := extractExpectedAudience(r)
+
+	// Normalize both audience values for RFC 3986 compliant comparison
+	normalizedTokenAudience := normalizeAudienceURI(token.Audience.String)
+	normalizedExpectedAudience := normalizeAudienceURI(expectedAudience)
+
+	// Validate that the token's audience matches the expected audience
+	if normalizedTokenAudience != normalizedExpectedAudience {
+		return xerrors.Errorf("token audience %q does not match expected audience %q",
+			token.Audience.String, expectedAudience)
+	}
+
+	return nil
+}
+
+// normalizeAudienceURI implements RFC 3986 URI normalization for OAuth2 audience comparison.
+// This ensures consistent audience matching between authorization and token validation.
+func normalizeAudienceURI(audienceURI string) string {
+	if audienceURI == "" {
+		return ""
+	}
+
+	u, err := url.Parse(audienceURI)
+	if err != nil {
+		// If parsing fails, return as-is to avoid breaking existing functionality
+		return audienceURI
+	}
+
+	// Apply RFC 3986 syntax-based normalization:
+
+	// 1. Scheme normalization - case-insensitive
+	u.Scheme = strings.ToLower(u.Scheme)
+
+	// 2. Host normalization - case-insensitive and IDN (punnycode) normalization
+	u.Host = normalizeHost(u.Host)
+
+	// 3. Remove default ports for HTTP/HTTPS
+	if (u.Scheme == "http" && strings.HasSuffix(u.Host, ":80")) ||
+		(u.Scheme == "https" && strings.HasSuffix(u.Host, ":443")) {
+		// Extract host without default port
+		if idx := strings.LastIndex(u.Host, ":"); idx > 0 {
+			u.Host = u.Host[:idx]
+		}
+	}
+
+	// 4. Path normalization including dot-segment removal (RFC 3986 Section 6.2.2.3)
+	u.Path = normalizePathSegments(u.Path)
+
+	// 5. Remove fragment (OAuth2 audience should not have fragments anyway)
+	u.Fragment = ""
```

Comment:
`u.Fragment` should be empty if we get here since we validate it beforehand. If we get here and `u.Fragment` is empty, something very weird has happened.